### ISSUE TITLE
[ppc64le pytorch] test_nn failures when using dtype=torch.float16

### ISSUE THREAD

**Issue by avmgithub** [pytorch/pytorch#7420]
I'm getting the following two test failures which may be related to using torch.float16
It only happens when running on ppc64le Power9 (volta gpu).    It does not happen on Power8 (pascal gpu)  More system details below.

FAIL: test_Conv2d_groups_nobias (__main__.TestNN)
 
Traceback (most recent call last):
  File "test_nn.py", line 3011, in test_Conv2d_groups_nobias
    dtype2prec[dtype])
  File "/home/freddie/pytorch/test/common.py", line 245, in assertEqual
    assertTensorsEqual(x, y)
  File "/home/freddie/pytorch/test/common.py", line 237, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: tensor(1.00000e-02 *
       1.5625, dtype=torch.float16, device='cuda:0') not less than or equal to 0.01 :

FAIL: test_variable_sequence_cuda (__main__.TestNN) (dtype=torch.float16)
 
Traceback (most recent call last):
  File "test_nn.py", line 65, in call_helper
    f(self, *args, dtype=dtype)
  File "test_nn.py", line 3348, in test_variable_sequence_cuda
    self._test_variable_sequence("cuda", dtype)
  File "test_nn.py", line 3340, in _test_variable_sequence
    self.assertEqual(p1.grad, p2.grad, dtype2prec[dtype])
  File "/home/freddie/pytorch/test/common.py", line 245, in assertEqual
    assertTensorsEqual(x, y)
  File "/home/freddie/pytorch/test/common.py", line 237, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: tensor(5.0391, dtype=torch.float16, device='cuda:0') not less than or equal to 0.01 :

ppc64le  
Power 9  4 x Volta GPU
RH7.4
Linux p95a28.pbm.ihost.com 4.11.0-44.5.1.el7a.ppc64le #1 SMP Thu Jan 18 02:18:03 EST 2018 ppc64le ppc64le ppc64le GNU/Linux
CUDA 9.1  CuDNN 7.0.5
torch (0.5.0a0+537cb10)

**Comment by avmgithub** [comment]
To further confirm my suspicion I pointed the torch.half test to cuda:1 as in 
dev_dtypes = [("cuda", torch.float), ("cuda:1", torch.float16)]
and the test passes

Is anyone in the community testing with cuda 9.1 , cudnn 7.0.5 with Volta GPUs?

**Comment by avmgithub** [comment]
For reference , I've opened an nvidia bug:  https://developer.nvidia.com/nvidia_bug/2115934

**Comment by ngimel** [comment]
The first error you are getting is a tolerance issue. It's a flaky test, so changing the order of the tests leads to different initialization, and the test passes. We've also seen the second error you are getting, but in our case the error was much smaller, so we assumed it is also a tolerance issue. In your case error of 5 doesn't look like tolerance, we'll look into it.

**Comment by csarofeen** [comment]
The second error looks like a valid 7.0.5 bug. The latest version of cuDNN we're running tests on is  7.1.2 and does not reproduce this error.

**Comment by avmgithub** [comment]
@ngimel @csarofeen I'm just tested cudnn 7.1.3 on a Power9 with cuda 9.1 and Volta GPU.  Here are the results for the second test :

$ python -m unittest test_nn.TestNN.test_variable_sequence_cuda -v
test_variable_sequence_cuda (test_nn.TestNN) ...

FAIL: test_variable_sequence_cuda (test_nn.TestNN) (dtype=torch.float16)

Traceback (most recent call last):
  File "/home/freddie/pytorch/test/test_nn.py", line 65, in call_helper
    f(self, *args, dtype=dtype)
  File "/home/freddie/pytorch/test/test_nn.py", line 3369, in test_variable_sequence_cuda
    self._test_variable_sequence("cuda", dtype)
  File "/home/freddie/pytorch/test/test_nn.py", line 3361, in _test_variable_sequence
    self.assertEqual(p1.grad, p2.grad, dtype2prec[dtype])
  File "/home/freddie/pytorch/test/common.py", line 245, in assertEqual
    assertTensorsEqual(x, y)
  File "/home/freddie/pytorch/test/common.py", line 230, in assertTensorsEqual
    self.assertTrue(torch.equal(nan_mask, b != b), message)
AssertionError: False is not true :

----------------------------------------------------------------------
Ran 1 test in 0.753s

For the first test: 

$ python -m unittest test_nn.TestNN.test_Conv2d_groups_nobias -v
test_Conv2d_groups_nobias (test_nn.TestNN) ... FAIL


FAIL: test_Conv2d_groups_nobias (test_nn.TestNN)
 
Traceback (most recent call last):
  File "/home/freddie/pytorch/test/test_nn.py", line 3032, in test_Conv2d_groups_nobias
    dtype2prec[dtype])
  File "/home/freddie/pytorch/test/common.py", line 245, in assertEqual
    assertTensorsEqual(x, y)
  File "/home/freddie/pytorch/test/common.py", line 237, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: tensor(1.00000e-02 *
       1.5625, dtype=torch.float16, device='cuda:0') not less than or equal to 0.01 :

----------------------------------------------------------------------
Ran 1 test in 0.612s

FAILED (failures=1)

**Comment by avmgithub** [comment]
After running python test_nn.py -v   I am getting the error for test_variable_sequence_cuda when using 7.1.3.   So it seems that the problem is not fixed on 7.1.3 for ppc64le

FAIL: test_variable_sequence_cuda (__main__.TestNN) (dtype=torch.float16)

Traceback (most recent call last):
  File "test_nn.py", line 65, in call_helper
    f(self, *args, dtype=dtype)
  File "test_nn.py", line 3369, in test_variable_sequence_cuda
    self._test_variable_sequence("cuda", dtype)
  File "test_nn.py", line 3361, in _test_variable_sequence
    self.assertEqual(p1.grad, p2.grad, dtype2prec[dtype])
  File "/home/freddie/pytorch/test/common.py", line 245, in assertEqual
    assertTensorsEqual(x, y)
  File "/home/freddie/pytorch/test/common.py", line 237, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: tensor(5.0391, dtype=torch.float16, device='cuda:0') not less than or equal to 0.01 :

----------------------------------------------------------------------
Ran 1124 tests in 644.076s

FAILED (failures=2, skipped=7)

**Comment by deepali-c** [comment]
I have observed similar failures:
```
$ python -m unittest test_nn.TestNN.test_variable_sequence_cuda -v
test_variable_sequence_cuda (test_nn.TestNN) ...
==================================================================
FAIL: test_variable_sequence_cuda (test_nn.TestNN) (dtype=torch.float16)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/builder/test/test_nn.py", line 65, in call_helper
    f(self, *args, dtype=dtype)
  File "/home/builder/test/test_nn.py", line 3270, in test_variable_sequence_cuda
    self._test_variable_sequence("cuda", dtype)
  File "/home/builder/test/test_nn.py", line 3260, in _test_variable_sequence
    self.assertEqual(x_leaf.grad, grad_x)
  File "/home/builder/test/common.py", line 243, in assertEqual
    assertTensorsEqual(x, y)
  File "/home/builder/test/common.py", line 235, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: tensor(1.00000e-05 *
       3.0518, dtype=torch.float16, device='cuda:0') not less than or equal to 1e-05 :

$ python -m unittest test_nn.TestNN.test_Conv2d_groups_nobias -v
test_Conv2d_groups_nobias (test_nn.TestNN) ... FAIL
==================================================================
FAIL: test_Conv2d_groups_nobias (test_nn.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/builder/test/test_nn.py", line 2933, in test_Conv2d_groups_nobias
    dtype2prec[dtype])
  File "/home/builder/test/common.py", line 243, in assertEqual
    assertTensorsEqual(x, y)
  File "/home/builder/test/common.py", line 235, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: tensor(1.00000e-02 *
       1.5625, dtype=torch.float16, device='cuda:0') not less than or equal to 0.01 :
```

System Details
```
ppc64le
Power9 - 4 x Tesla GPU
RH7.5
CUDA 9.2 CuDNN 7.1.4
torch 0.4.0
```

**Comment by hartb** [comment]
I'm looking more at `test_variable_sequence_cuda`. It fails consistently in the environment listed just above--always with:

`AssertionError: tensor(1.00000e-05 *
       3.0518, dtype=torch.float16, device='cuda:0') not less than or equal to 1e-05`


The test fails only for float16.  In the failing case, the mismatch is in a single member of the tensor:

```
$ diff x_leaf.grad grad_x 
4c4
<          [-0.0182,  0.0000,  0.0499],
---
>          [-0.0183,  0.0000,  0.0499],
```

And the difference is always `3.0518e-05` (which happens == `2**-16`).

I notice the assert here is coded as:

`self.assertEqual(x_leaf.grad, grad_x)`

Which demands a precision of 1e-05.  But several of the other `assertEqual()` in this file were adjusted (by: https://github.com/pytorch/pytorch/commit/1c01eabd3cc8402e6de815245774278af3ad0c7f) to relax the precision to `1e-02` for float16, including one just below in this same test:

```
dtype2prec = {torch.float: 1e-5,
              torch.double: 1e-5,
              torch.half: 1e-2}
...
            self.assertEqual(p1.grad, p2.grad, dtype2prec[dtype])
```

Is this maybe a case of the test tolerance being too tight?  Or should we really expect the first comparison to fall within `1e-05`, even for float16?

**Comment by hartb** [comment]
If I test only float16 the results change a bit, but are still repeatable, now:

```
AssertionError: tensor(1.00000e-05 *     
       1.5258789062, dtype=torch.float16, device='cuda:0') not less than or equal to 1e-05
```

and

```
$ diff x_leaf.grad grad_x                               
5c5                                      
<          [0.0141143799, 0.0502929688, -0.0027675629],                           
---                                      
>          [0.0140991211, 0.0502929688, -0.0027675629],
```

And the miss-match is now always `2**-16` (I misspoke above, `3.0518e-05` == `2**-15`).

**Comment by hartb** [comment]
Looking at the failure of `test_Conv2d_groups_nobias`...

This is also a "value out of tolerance" failure that we see only when using CUDA and torch.half/float16 on a system with V100 GPUs.

The failure is data specific. I cut down the test script to include just the failing case with fixed parameters and input values. (See attached `Conv2d.py`.)

With that setup, the values that differ out of tolerance are weights. The failing assert is:

```
        self.assertEqual(m.weight.grad.data[2:], m2.weight.grad.data, 1e-2)
```

And the differing values:

```
$ python Conv2d.py
tensor([[[[-0.23413086, 16.67187500, -12.15625000],
...                     ^^^^^^^^^^^
tensor([[[[-0.23486328, 16.68750000, -12.15625000],
...                     ^^^^^^^^^^^
AssertionError: tensor(1.00000e-02 *
       1.56250000, dtype=torch.float16, device='cuda:0') not less than or equal to 0.01
```

Note the mis-matched values are beween 16 and 32. In that range the limit of precision for float16 should be `2^-6 == 1.5625e-02` (at least per: https://en.wikipedia.org/wiki/Half-precision_floating-point_format), which is exactly the mismatch difference. So the values mismatch by a single bit at the limit of resolution for float16 in this range, and the tolerance of the assert is finer than the resolution of float16 in this range.

So if the test intends to allow slight imprecision at the resolution of the type, then the assert for the weights may be too strict. But if calculation should be deterministic and values exactly match, then there's something deeper gone wrong.

**Comment by hartb** [comment]
[Conv2d.zip](https://github.com/pytorch/pytorch/files/2150108/Conv2d.zip)

**Comment by avmgithub** [comment]
fixed by #10519 , closing
